pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -0,0 +1,146 @@
+type StatefulSetControlInterface interface {
+	// UpdateStatefulSet implements the control logic for Pod creation, update, and deletion, and
+	// persistent volume creation, update, and deletion.
+	// If an implementation returns a non-nil error, the invocation will be retried using a rate-limited strategy.
+	// Implementors should sink any errors that they do not wish to trigger a retry, and they may feel free to
+	// exit exceptionally at any point provided they wish the update to be re-run at a later point in time.
+	UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error
+}
+
+// NewDefaultStatefulSetControl returns a new instance of the default implementation StatefulSetControlInterface that
+// implements the documented semantics for StatefulSets. podControl is the PodControlInterface used to create, update,
+// and delete Pods and to create PersistentVolumeClaims. You should use an instance returned from
+// NewRealStatefulPodControl() for any scenario other than testing.
+func NewDefaultStatefulSetControl(podControl StatefulPodControlInterface) StatefulSetControlInterface {
+	return &defaultStatefulSetControl{podControl}
+}
+
+type defaultStatefulSetControl struct {
+	podControl StatefulPodControlInterface
+}
+
+func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {
+	replicaCount := int(*set.Spec.Replicas)
+	replicas := make([]*v1.Pod, replicaCount)
+	condemned := make([]*v1.Pod, 0, len(pods))
+	ready := 0
+	unhealthy := 0
+
+	// First we partition pods into two lists, replicas will contain all Pods such that
+	// 0 <= getOrdinal(pod) < set.Spec.Replicas and condemned will contain all Pods such that
+	// set.Spec.Replicas <= getOrdinal(pod)
+	for i := range pods {
+		//count the number of running and ready replicas
+		if isRunningAndReady(pods[i]) {
+			ready++
+		}
+		if ord := getOrdinal(pods[i]); 0 <= ord && ord < replicaCount {
+			// if the ordinal of the pod is within the range of the current number of replicas,
+			// insert it at the indirection of its ordinal
+			replicas[ord] = pods[i]
+
+		} else if ord >= replicaCount {
+			// if the ordinal is greater than the number of replicas add it to the condemned list
+			condemned = append(condemned, pods[i])
+		}
+	}
+
+	// for any empty indices in the sequence [0,set.Spec.Replicas) create a new Pod
+	for ord := 0; ord < replicaCount; ord++ {
+		if replicas[ord] == nil {
+			replicas[ord] = newStatefulSetPod(set, ord)
+		}
+	}
+
+	// count the number of unhealthy pods
+	for i := range replicas {
+		if !isHealthy(replicas[i]) {
+			unhealthy++
+		}
+	}
+	for i := range condemned {
+		if !isHealthy(condemned[i]) {
+			unhealthy++
+		}
+	}
+
+	// sort the condemned Pods by their ordinals
+	sort.Sort(ascendingOrdinal(condemned))
+
+	// attempt to update set with the current number of replicas
+	set.Status = apps.StatefulSetStatus{Replicas: int32(ready)}
```

Comment:
I don't think this is right because you end up wiping out all the other fields of the ss status. You should contstruct the new status based on the current one, update the fields you need to update, and then compare the two statuses and do the actual update if they don't match. FWIW, [here](https://github.com/kubernetes/kubernetes/blob/2541c16692c7777b0aeda8124f7895855b1b9232/pkg/controller/deployment/sync.go#L569) is what we are doing for Deployments.